### PR TITLE
ARROW-3002: [Python] Hash more parts of pyarrow.Field

### PR DESCRIPTION
**Pull request by kszucs**

### CODE REVIEW

File: python/pyarrow/types.pxi [hunk 1]
```diff
@@ -420,7 +420,7 @@ cdef class Field:
         return self.__str__()
 
     def __hash__(self):
-        return hash((self.field.name(), self.type.id))
+        return hash((self.field.name(), self.type, self.field.nullable()))
```
Comment (kszucs) [inline python/pyarrow/types.pxi]: I guess We should hash the metadata too. 
The hashing could be factored out, it's equivalent with

```py
hash(self.__reduce__()[1])
```

Comment (pitrou) [reply python/pyarrow/types.pxi]: IMO the current form is preferred. The `__reduce__` variant is cryptic.
And no, I don't think we should hash the metadata. A hash function should be fast while minimizing _real-world_ collisions.

File: python/pyarrow/tests/test_types.py [hunk 2]
```diff
@@ -44,10 +44,15 @@
...
+               pa.field('b', pa.int8(), nullable=False),
+               pa.field('c', pa.string())]),
     pa.union([pa.field('a', pa.binary(10)),
               pa.field('b', pa.string())], mode=pa.lib.UnionMode_DENSE),
     pa.union([pa.field('a', pa.binary(10)),
               pa.field('b', pa.string())], mode=pa.lib.UnionMode_SPARSE),
+    pa.union([pa.field('a', pa.binary(10), nullable=False),
+              pa.field('b', pa.string())], mode=pa.lib.UnionMode_SPARSE),
     # XXX Needs array pickling
     # pa.dictionary(pa.int32(), pa.array(['a', 'b', 'c'])),
```
Comment (pitrou) [inline python/pyarrow/tests/test_types.py]: Unrelated, but what happens if you uncomment this one? Array pickling should work now.

Comment (kszucs) [reply python/pyarrow/tests/test_types.py]: Trying that

File: python/pyarrow/types.pxi (hunk 1 above)
Comment (kszucs) [reply python/pyarrow/types.pxi]: Great explanation, thanks!

**Comment by kszucs** [comment]
`__reduce__` was missing for dictionary type, but it works now

**Comment by pitrou** [comment]
Hmm, looks like the added test messes the test for `total_allocated_bytes` (see AppVeyor failure).

**Comment by kszucs** [comment]
Does `total_allocated_bytes` test assume, that no arrow array has been allocated before that test case? 

Perhaps because `MANY_TYPES` is a module level variable, whic holds a ref for an array.

I can create a parametrized datatype fixture instead of a global list, which would probably keep total_allocated_bytes as zero. Is that the desired behaviour?

**Comment by wesm** [comment]
I wonder where the memory allocation is happening. It's possible there is a memory leak somewhere or a dangling PyObject* that has not been gc'd yet (or has leaked)

**Comment by pitrou** [comment]
> I can create a parametrized datatype fixture instead of a global list,

Or you could simply write a helper function that returns that list.

**Comment by kszucs** [comment]
But then We can't use it for parametrization.

**Comment by pitrou** [comment]
As you prefer. I generally feel that parametrization is overrated, though it can give more informative errors.

**Comment by kszucs** [comment]
I can admit it has its downsides.

**Comment by wesm** [comment]
I personally don't care for `pytest.parametrize`. Among other things, it can give a false sense of a test suite being much larger than it is (I've commented to @cpcloud about this several times). It's also a bit of a bummer that parametrize does not support fixtures as arguments

**Comment by cpcloud** [comment]
I like parametrize.

The body of a test IMO should contain the "kernel" of what's being tested and a loop that's iterating over a bunch test cases is not part of that kernel. I also want to be able to specify what specific value of a parameter I can test.

parametrize was pretty useful in writing [a set of tests for ibis](https://github.com/ibis-project/ibis/tree/master/ibis/tests/all) that run a subset of tests for across each backend.

i would've wasted a lot of time and the code would've been less readable IMO if i had to write the loops by hand, and in fact i would've just been rewriting machinery that already existed.

I see no reason not to make use of tool because someone else is able to convince themselves that they are being more productive by seeing more dots on a screen. The merit of a tool shouldn't be based on how others misuse it.

As far as passing fixtures to parametrize goes, I think that probably just needs a bit of old-fashioned open source TLC: https://github.com/pytest-dev/pytest/issues/349.

**Comment by kszucs** [comment]
Actually that's possible with [indirect](https://docs.pytest.org/en/latest/example/parametrize.html#apply-indirect-on-particular-arguments).

Whatever, if We want to test that `total_allocated_bytes == 0` We can't use parametrization, a ref would be stored unless we parametrize with lambdas - which we obviously don't want to :)

BTW I think `test_array.py::test_total_bytes_allocated` is too sensitive for side effects.

**Comment by cpcloud** [comment]
Though I'm happy to defer to the existing style in a codebase, I'm not a zealot :)

**Comment by cpcloud** [comment]
Ah, I forgot about `indirect`. It's definitely a bit brain-melting at first.

**Comment by kszucs** [comment]
> Ah, I forgot about indirect. It's definitely a bit brain-melting at first.
I can agree with that

**Comment by pitrou** [comment]
> i would've wasted a lot of time and the code would've been less readable IMO if i had to write the loops by hand, and in fact i would've just been rewriting machinery that already existed.

This is what I don't understand. Writing a loop isn't more costly than writing a decorator...
And the idea that we shouldn't use loops in tests while we use them all the time in regular code sounds a bit weird to me.

**Comment by cpcloud** [comment]
> Writing a loop isn't more costly than writing a decorator...

What is the cost function here? What about when you want to parallelize your tests? This is much more tractable if the thing parallelizing your tests also controls the case iteration.

Ultimately this is a religious debate and probably counterproductive. Let's agree to disagree here.

**Comment by wesm** [comment]
> Writing a loop isn't more costly than writing a decorator...

I'm of @pitrou's mindset on this aspect, but I'm not going to make anyone change any code. I would rather write loops and put the "kernel" of the test function in a helper function (like the `_check_$THING` functions I'm so fond of writing) if it helps readability or code reuse. Refactoring plain old Python is more pleasant (to me) than Python + some parameter injection magic from pytest. 

In any case, I see no point to be dogmatic. I'm generally going to write loops in my tests over parametrize

**Comment by wesm** [comment]
The auto-parallelization point is well taken

File: python/pyarrow/tests/test_types.py [hunk 3]
```diff
@@ -285,16 +296,18 @@ def test_decimal_byte_width():
...
 
-    # could use two parametrization levels, but that'd bloat pytest's output
-    for i, other in enumerate(MANY_TYPES + non_pyarrow):
-        if i == index:
-            assert ty == other
-        else:
-            assert ty != other
+    for index, ty in enumerate(many_types):
+        # could use two parametrization levels,
+        # but that'd bloat pytest's output
```
Comment (pitrou) [inline python/pyarrow/tests/test_types.py]: Ha :-)

**Review by pitrou — APPROVED** [review]
LGTM. Thanks @kszucs !

**Pull request closed without merging**